app/src/ui/main-process-proxy.ts

```diff
@@ -74,16 +74,25 @@ let currentContextualMenuItems: ReadonlyArray<IMenuItem> | null = null
 export function registerContextualMenuActionDispatcher() {
   ipcRenderer.on(
     'contextual-menu-action',
-    (event: Electron.IpcRendererEvent, index: number) => {
-      if (!currentContextualMenuItems) {
+    (event: Electron.IpcRendererEvent, indices: number[]) => {
+      if (currentContextualMenuItems == null) {
         return
       }
-      if (index >= currentContextualMenuItems.length) {
-        return
+
+      let foundMenuItem: IMenuItem = {
```

Comment:
This is one of the places where TypeScript not only won't have our backs, it'll even stab us in it when we're not looking.

As a reader of this code I look at this statement and think cool, foundMenuItem will never be `undefined`  or `null`. But that's not necessarily true since `foundMenuItem.submenu[index]` doesn't come with any guarantees that the index we're looking for exists even though the type system fools us into believing that.

So in a worst-case scenario here where the `currentContextualMenuItems` global variable ends up gets swapped at an inopportune time we _could_ (most likely we wouldn't though) end up with `foundMenuItem` that's `undefined` and :boom:

Given that we don't anticipate ever having an unbounded depth of submenus (because that'd be 🍌:banana:) I wonder if we can both categorically guarantee that it won't be undefined and at the same time uphold out [immutability](https://github.com/desktop/desktop/blob/development/docs/contributing/engineering-values.md#immutability-is-good) preference by adding a recursive helper function. Something like this perhaps

```ts
const foundMenuItem = findSubmenuItem(
  { submenu: currentContextualMenuItems },
  indices
)

if (foundMenuItem !== undefined && foundMenuItem.action !== undefined) {
  foundMenuItem.action()
  currentContextualMenuItems = null
}
```

```ts
/**
 * Recursively traverse the submenu(s) of the given menu item following
 * the indices provided until the desired submenu item could be found.
 *
 * Returns undefined if menu could not be traverse in accordance with
 * the indices parameter.
 */
function findSubmenuItem(
  menuItem: IMenuItem | undefined,
  indices: ReadonlyArray<number>,
  depth: number = 0
): IMenuItem | undefined {
  if (
    menuItem === undefined ||
    menuItem.submenu === undefined ||
    depth >= indices.length
  ) {
    return undefined
  }

  return depth === indices.length - 1
    ? menuItem.submenu[indices[depth]]
    : findSubmenuItem(menuItem.submenu[indices[depth]], indices, depth + 1)
}
```

Reply:
Good catch! You concern is real though and I think it's quite possible that this could happen under some race condition (although it would be a very rare situation). 

This makes we wonder whether apart from fixing this specific `undefined` issue we should think about adding some logic to warranty that the `currentContextualMenuItems` object doesn't change between the time when the context menu gets opened and the time that an item gets clicked: e.g it would also be very bad if a user clicks on a tag to delete it and we end up deleting a different one because a new tag appears on the list at the appropiate moment.

Regarding your proposed solution, I'm not convinced about changing a simple for loop by a recursive function (I'd rather add an `if` check when accessing the array index), since making the function recursive doesn't solve the root problem of TypeScript [not returning `undefined` for index lookups](https://github.com/microsoft/TypeScript/issues/13778) and imho makes the logic harder to reason.

I agree about prefering immutabilty over mutabilty (and requiring it on structures that are shared between functions), but I think having a local mutation sometimes is a more pragmatic approach (see: https://twitter.com/dan_abramov/status/1099777642885189633).

Maybe I can move the `for` loop to its own method so the variable with a mutable assignment has less scope (even if the original method grows in size)? something like:

```ts
function findSubmenuItem(
  currentContextualMenuItems: ReadonlyArray<IMenuItem>,
  indices: ReadonlyArray<number>
): IMenuItem | undefined {
  let foundMenuItem: IMenuItem | undefined = {
    submenu: currentContextualMenuItems,
  }

  // Traverse the submenus of the context menu until we find the appropiate index.
  for (const index of indices) {
    if (foundMenuItem === undefined || foundMenuItem.submenu === undefined) {
      return undefined
    }
    foundMenuItem = foundMenuItem.submenu[index]
  }

  return foundMenuItem
}
```

Reply:
> (I'd rather fix the type of foundMenuItem and add an additional if check

Yeah, that was my first thought as well but TS type narrowing bites us in the behind here as it'll always infer the type of `foundMenuItem` to be `IMenuItem` since the last executable statement of the `for` narrows it to `IMenuItem`. So even explicitly typing the variable as `IMenuItem | undefined` won't afford us any compile-time guarantees in case someone were to attempt to remove the added `if (foundMenuItem !== undefined)` check.

> Maybe I can move the for loop to its own method so the variable with a mutable assignment has less scope?

Yes, I think that's a lovely compromise!

Reply:
>  it'll always infer the type of foundMenuItem to be IMenuItem since the last executable statement of the for narrows it to IMenuItem

Luckily in this specific case, since the block inside a for loop is not guaranteed to be executed (e.g empty array) TS can't narrow the type, but I agree that this could bite us in any other scenario... Unfortunately this is something I don't expect in the short-mid term to get fixed in TS since it would cause tons of errors on existing apps 😟